### PR TITLE
Base for Venue and Organizer REST API insertion support

### PR DESCRIPTION
**Pull request by lucatume**
Tickets: https://central.tri.be/issues/79740 and https://central.tri.be/issues/79737

Used in https://github.com/moderntribe/the-events-calendar/pull/1423

This PR adds the code needed to support REST API insertion of Venues and Organizers; I've added the `Tribe__Data` class too to have something we can reuse to juggle maybe-defined data and so on.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Expected 1 space after closing parenthesis; found 0 | [Data.php:126](https://github.com/moderntribe/tribe-common/blob/a0ba90875e3782aba7fe7f11bd23fe25acae1453/src/Tribe/Data.php#L126)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Avoid function calls in a FOR loop test part | [Post.php:149](https://github.com/moderntribe/tribe-common/blob/2c9e8d20d713939b5db67d3f0312cf8ebb3af25c/src/Tribe/Duplicate/Post.php#L149)
:x: | Expected 1 space after closing parenthesis; found 0 | [Like.php:43](https://github.com/moderntribe/tribe-common/blob/2c9e8d20d713939b5db67d3f0312cf8ebb3af25c/src/Tribe/Duplicate/Strategy/Like.php#L43)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
Ok, I've added here the base for the duplicate finding mechanics (last two commits); still no code in TEC but that's the base that will be used.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Expected 1 space after closing parenthesis; found 0 | [Strategy_Factory.php:60](https://github.com/moderntribe/tribe-common/blob/3bf5a3512ce4e5ccb5364c32aacce6570e8e2cff/src/Tribe/Duplicate/Strategy_Factory.php#L60)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Duplicate/Strategy/Like.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		if ( $this->is_a_numeric_post_field( $key ) ) {
+			return $wpdb->prepare( "{$key} = %d", $value );
+		}
+
+		$frags = $this->get_where_frags( $value );
+
+		$where_frags = array();
+		foreach ( $frags as $frag ) {
+			$prepared = $wpdb->prepare( '%s', '%' . strtolower( trim( $frag ) ) . '%' );
+			$where_frags[] = "{$key} LIKE {$prepared}";
```
Comment (sc0ttkclark) [inline src/Tribe/Duplicate/Strategy/Like.php]: This reads a little odd, maybe these two lines should be more like:

```php
$formatted_frag = '%' . strtolower( trim( $frag ) ) . '%';
$where_frags[] = $wpdb->prepare( "{$key} LIKE %s", $formatted_frag );
```

I'm also wondering if these need to get run through `$wpdb->esc_like()` too?

```php
$formatted_frag = '%' . $wpdb->esc_like( strtolower( trim( $frag ) ) ) . '%';
```

File: src/Tribe/Duplicate/Strategy/Like.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	public function where_custom_field( $key, $value, $table_alias ) {
+		/** @var wpdb $wpdb */
+		global $wpdb;
+
+		$frags = $this->get_where_frags( $value );
+
+		$where_frags = array();
+		foreach ( $frags as $frag ) {
+			$meta_value = $wpdb->prepare( '%s', '%' . strtolower( trim( $frag ) ) . '%' );
+			$where_frags[] = $wpdb->prepare( "{$table_alias}.meta_key = %s AND {$table_alias}.meta_value LIKE ", $key ) . $meta_value;
```
Comment (sc0ttkclark) [inline src/Tribe/Duplicate/Strategy/Like.php]: This reads a little odd, maybe these two lines should be more like:

```php
$formatted_frag = '%' . strtolower( trim( $frag ) ) . '%';
$where_frags[] = $wpdb->prepare( "{$table_alias}.meta_key = %s AND {$table_alias}.meta_value LIKE %s", $key, $formatted_frag );
```

I'm also wondering if these need to get run through `$wpdb->esc_like()` too?

```php
$formatted_frag = '%' . $wpdb->esc_like( strtolower( trim( $frag ) ) ) . '%';
```

Comment (lucatume) [inline src/Tribe/Duplicate/Strategy/Like.php]: This and the other one both good ideas: thanks.
I've modified  the code.

**Comment by lucatume** [comment]
The work is complete here and good for a review.

File: src/Tribe/Data.php [hunk 3]
```diff
@@ -0,0 +1,224 @@
+<?php
+
+/**
+ * Class Tribe__Data
```
Comment (borkweb) [inline src/Tribe/Data.php]: Nice! This is a solid idea!

File: src/Tribe/REST/Endpoints/POST_Endpoint_Interface.php [hunk 4]
```diff
@@ -19,4 +19,9 @@ public function post( WP_REST_Request $request, $return_id = false );
 	 * @return array
 	 */
 	public function POST_args();
+
+	/**
+	 * @return bool Whether the current user can post or not.
+	 */
+	public function can_post(  );
```
Comment (borkweb) [inline src/Tribe/REST/Endpoints/POST_Endpoint_Interface.php]: Remove the two spaces from inside the parentheses.

**Pull request merged**